python/tvm/relay/ir_pass.py

```diff
@@ -1,16 +1,369 @@
+
+        Parameters
+        ----------
+        mod : tvm.relay.Module
+            The module that a certain optimization is performed on.
+
+        Returns
+        -------
+        mod : tvm.relay.Module
+            The updated module after applying this pass.
+        """
+        return _ir_pass.RunPass(self, mod)
+
+
+@register_relay_node
+class ModulePass(Pass):
+    """A pass that works on tvm.relay.Module.
+
+    Parameters
+    ----------
+    name : str
+        The module pass name.
+
+    opt_level : int
+        The optimization level of this module pass.
+
+    required : List[str]
+        The list of passes that are required by a module pass.
+
+    pass_func : Callable[(tvm.relay.Module, PassContext) -> tvm.relay.Module]
+        The callback function that sketches a certain optimization.
+    """
+
+    def __init__(self, name, opt_level, pass_func, required=None):
+        self.__init_handle_by_constructor__(_ir_pass.CreateModulePass,
+                                            name, opt_level, required,
+                                            pass_func)
+
+
+@register_relay_node
+class FunctionPass(Pass):
+    """A pass that works on each tvm.relay.Function in a module.
+
+    Parameters
+    ----------
+    name : str
+        The function pass name.
+
+    opt_level : int
+        The optimization level of this function pass.
+
+    required : List[str]
+        The list of passes that are required by a function pass.
+
+    pass_func : Callable[(tvm.relay.Function, PassContext) ->
+                tvm.relay.Function]
+        The callback function that sketches a certain optimization.
+    """
+
+    def __init__(self, name, opt_level, pass_func, required=None):
+        self.__init_handle_by_constructor__(_ir_pass.CreateFunctionPass,
+                                            name, opt_level, required,
+                                            pass_func)
+
+
+@register_relay_node
+class SequentialPass(Pass):
+    """A pass that works on a sequence of pass objects.
+
+    Parameters
+    ----------
+    name : str
+        The sequential pass name.
+
+    opt_level : int
+        The optimization level of this sequential pass.
+
+    passes : List[Pass]
+        The pass candidates to be executed.
+
+    required : List[str]
+        The list of passes that are required by a sequential pass.
+
+    disabled : Optional[List[str]]
+        The list of passes that are disabled.
+    """
+
+    def __init__(self, name, opt_level, passes, required=None, disabled=None):
+        disabled = disabled if disabled else []
+        if not isinstance(disabled, (list, tuple)):
+            raise TypeError("disabled must be a list or tuple of pass names")
+        self.__init_handle_by_constructor__(_ir_pass.CreateSequentialPass,
+                                            name, opt_level, passes, required,
+                                            disabled)
+
+
+def module_pass(opt_level, name=None, required=None):
```

Comment:
Consistency in the convention:

Let us make the argument list as follows:

```python
def module_pass(pass_func=None, opt_level=None, name=None, required=None):
    """When pass_func == None, decorator mod, otherwise normal mode.
    """     
```

Please also change all the constructors to keep the consistent order of arguments

Reply:
API design is really important if we take a look at constructor order and the function order they are different. 
This can be quite problematic and confuses user in the long run :)

Reply:
Thanks. Using this signature, now I think we should remove those initialization functions.